Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h

```diff
@@ -545,12 +529,17 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
 #endif
 
     String migrateMediaKeysStorageIfNecessary(const String& directory);
+    void dispatchResolveDirectories();
+    void resolveDirectories(WebsiteDataStoreConfiguration::Directories&);
 
     const PAL::SessionID m_sessionID;
 
-    Ref<WebsiteDataStoreConfiguration> m_resolvedConfiguration;
+    Lock m_resolveLock;
+    Condition m_resolveCondition;
+    enum class ResolveState : uint8_t { Unresolved, Resolving, Resolved };
+    ResolveState m_resolveState WTF_GUARDED_BY_LOCK(m_resolveLock) { ResolveState::Unresolved };
+    RefPtr<WebsiteDataStoreConfiguration> m_resolvedConfiguration WTF_GUARDED_BY_LOCK(m_resolveLock);
```

Comment:
It is not clear to me which we still need a `m_resolvedConfiguration` data member with your change. Why don't we just have a `Directories m_resolvedDirectories` data member? It would also look better in terms of thread safety. Right now you have a lock on the `m_resolvedConfiguration` data member. But if someone somewhere outside WebsiteDataStore uses/modifies it, access won't be protected by the lock. Having a local struct of Strings instead of a RefPtr is much clearer for thread safety.

Reply:
Maybe it can even be a `std::optional<Directories>` and then we can get rid of `ResolveState`? I don't really think we truly need 3 states?